rotkehlchen/chain/evm/decoding/paraswap/decoder.py

```diff
@@ -0,0 +1,190 @@
+from rotkehlchen.chain.evm.decoding.interfaces import DecoderInterface
+from rotkehlchen.chain.evm.decoding.structures import (
+    DEFAULT_DECODING_OUTPUT,
+    DecoderContext,
+    DecodingOutput,
+)
+from rotkehlchen.chain.evm.decoding.types import CounterpartyDetails
+from rotkehlchen.chain.evm.decoding.utils import maybe_reshuffle_events
+from rotkehlchen.history.events.structures.types import HistoryEventSubType, HistoryEventType
+from rotkehlchen.types import ChecksumEvmAddress
+from rotkehlchen.utils.misc import hex_or_bytes_to_address, hex_or_bytes_to_int
+
+from .constants import BUY_SIGNATURE, CPT_PARASWAP, SWAP_SIGNATURE as PARASWAP_SWAP_SIGNATURE
+
+if TYPE_CHECKING:
+    from rotkehlchen.chain.evm.decoding.base import BaseDecoderTools
+    from rotkehlchen.chain.evm.node_inquirer import EvmNodeInquirer
+    from rotkehlchen.user_messages import MessagesAggregator
+
+
+class ParaswapCommonDecoder(DecoderInterface):
+    def __init__(
+            self,
+            evm_inquirer: 'EvmNodeInquirer',
+            base_tools: 'BaseDecoderTools',
+            msg_aggregator: 'MessagesAggregator',
+            router_address: ChecksumEvmAddress,
+            fee_receiver_address: ChecksumEvmAddress,
+    ) -> None:
+        super().__init__(evm_inquirer, base_tools, msg_aggregator)
+        self.router_address = router_address
+        self.fee_receiver_address = fee_receiver_address
+
+    def _decode_swap(
+            self,
+            context: DecoderContext,
+            receiver: ChecksumEvmAddress,
+            sender: ChecksumEvmAddress,
+    ) -> DecodingOutput:
+        """This function is used to decode the swap done by paraswap.
+        At the moment it does not decode the fee in ETH. For example:
+        etherscan.io/tx/0xeeea20b39f157fe59fa4904fd4b62f8971188b53d05c6831e0ed67ee157e40c2"""
+        if not self.base.is_tracked(sender) and not self.base.is_tracked(receiver):
+            return DEFAULT_DECODING_OUTPUT
+
+        out_event = in_event = returned_event = None
+        for event in context.decoded_events:
+            if (
+                event.event_type == HistoryEventType.TRADE and
+                event.event_subtype == HistoryEventSubType.SPEND or
+                event.event_type == HistoryEventType.SPEND and
+                event.event_subtype == HistoryEventSubType.NONE and
+                event.location_label == sender
+            ):  # find the send event
+                event.counterparty = CPT_PARASWAP
+                event.event_type = HistoryEventType.TRADE
+                event.event_subtype = HistoryEventSubType.SPEND
+                event.notes = f'Swap {event.balance.amount} {event.asset.resolve_to_asset_with_symbol().symbol} in paraswap'  # noqa: E501
+                event.address = self.router_address
+                out_event = event
+            elif event.event_type == HistoryEventType.RECEIVE and event.location_label == sender:  # find the receive event  # noqa: E501
```

Comment:
but did you see other transactions? or it is just this one that has some special conditions and doesn't get properly decoded by other decoders?

Reply:
As for the discussion in discord I believe is a bug to not cover the cases of trade/receive and not using the subtype check here for receive none. You are adding checks bassed on tests and not the information that you have. Since  a decoder could decode all the transaction you need to cover that case. That is why it is important to know what is happening